packages/devtools_app/lib/src/debugger/program_explorer.dart

```diff
@@ -0,0 +1,407 @@
+        decoration: icon == null
+            ? BoxDecoration(
+                color: color,
+                shape: BoxShape.circle,
+              )
+            : null,
+        child: icon == null
+            ? Center(
+                child: Text(
+                  character,
+                  style: TextStyle(
+                    height: 1,
+                    fontFamily: theme.fixedFontStyle.fontFamily,
+                    color: theme.colorScheme.defaultBackgroundColor,
+                    fontSize: chartFontSizeSmall,
+                  ),
+                  // Required to center the individual character within the
+                  // shape. Since letters like 'm' are shorter than letters
+                  // like 'f', there's padding applied to the top of shorter
+                  // characters in order for everything to align properly.
+                  // Since we're only dealing with individual characters, we
+                  // want to disable this behavior so shorter characters don't
+                  // appear to be slightly below center.
+                  textHeightBehavior: TextHeightBehavior(
+                    applyHeightToFirstAscent: isShortCharacter,
+                    applyHeightToLastDescent: false,
+                  ),
+                ),
+              )
+            : Icon(
+                icon,
+                size: defaultIconSize,
+                color: color,
+              ),
+      ),
+    );
+  }
+}
+
+/// Picker that displays the program's structure, allowing for navigation and
+/// filtering.
+class ProgramExplorer extends StatelessWidget {
+  ProgramExplorer({
+    Key key,
+    @required this.debugController,
+    @required this.onSelected,
+  }) : controller = programExplorerController, super(key: key) {
+    controller.selectScriptNode(debugController.currentScriptRef.value);
+  }
+
+  final ProgramExplorerController controller;
+  final DebuggerController debugController;
+  final void Function(ScriptLocation) onSelected;
+
+  @override
+  Widget build(BuildContext context) {
+    return ValueListenableBuilder<bool>(
+      valueListenable: controller.initialized,
+      builder: (context, initialized, _) {
+        Widget body;
+        if (!initialized) {
+          body = const CenteredCircularProgressIndicator();
+        } else {
+          body = LayoutBuilder(
+            builder: (context, constraints) {
+              return FlexSplitColumn(
+                totalHeight: constraints.maxHeight,
+                initialFractions: const [0.7, 0.3],
+                minSizes: const [0.0, 0.0],
+                headers: const <PreferredSizeWidget>[
+                  AreaPaneHeader(
+                    title: Text('File Explorer'),
+                    needsTopBorder: false,
+                  ),
+                  AreaPaneHeader(title: Text('Outline')),
+                ],
+                children: [
+                  ValueListenableBuilder<List<VMServiceObjectNode>>(
```

Comment:
This build method is deep. Instead, create 2 stateless widgets `FileExplorer` and `FileOutline` or similar, and use those widgets here.